### PR TITLE
Rewrite memory-charging feature's option API

### PR DESCRIPTION
**Pull request by hx235**
**Context:**
Previous PR https://github.com/facebook/rocksdb/pull/9748, https://github.com/facebook/rocksdb/pull/9073, https://github.com/facebook/rocksdb/pull/8428 added separate flag for each charged memory area. Such API design is not scalable as we charge more and more memory areas. Also, we foresee an opportunity to consolidate this feature with other cache usage related features such as `cache_index_and_filter_blocks` using `CacheEntryRole`.  

Therefore we decided to consolidate all these flags with `CacheUsageOptions cache_usage_options` and this PR serves as the first step by consolidating memory-charging related flags.

**Summary:**
- Replaced old API reference with new ones, including making `kCompressionDictionaryBuildingBuffer` opt-out and added a unit test for that
- Added missing db bench/stress test for some memory charging features
- Renamed related test suite to indicate they are under the same theme of memory charging
- Refactored a commonly used mocked cache component in memory charging related tests to reduce code duplication
- Replaced the phrases "memory tracking" / "cache reservation" (other than CacheReservationManager-related ones) with "memory charging" for standard description of this feature.

**Test plan:**
- New unit test for opt-out `kCompressionDictionaryBuildingBuffer` `TEST_F(ChargeCompressionDictionaryBuildingBufferTest, Basic)`
- New unit test for option validation/sanitization `TEST_F(CacheUsageOptionsOverridesTest, SanitizeAndValidateOptions)`
- CI
- db bench (in case querying new options introduces regression) **+0.5% micros/op**: `TEST_TMPDIR=/dev/shm/testdb ./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR  -charge_compression_dictionary_building_buffer=1(remove this for comparison)  -compression_max_dict_bytes=10000 -disable_auto_compactions=1 -write_buffer_size=100000 -num=4000000 | egrep 'fillseq'`

#-run | (pre-PR) avg micros/op | std micros/op | (post-PR)  micros/op | std micros/op | change (%)
-- | -- | -- | -- | -- | --
10 | 3.9711 | 0.264408 | 3.9914 | 0.254563 | 0.5111933721
20 | 3.83905 | 0.0664488 | 3.8251 | 0.0695456 | **-0.3633711465**
40 | 3.86625 | 0.136669 | 3.8867 | 0.143765 | **0.5289363078**

- db_stress: `python3 tools/db_crashtest.py blackbox  -charge_compression_dictionary_building_buffer=1 -charge_filter_construction=1 -charge_table_reader=1 -cache_size=1` killed as normal

### CODE REVIEW

File: table/block_based/block_based_table_factory.cc [hunk 1]
```diff
@@ -234,6 +234,7 @@ static std::unordered_map<std::string, OptionTypeInfo>
         /* currently not supported
           std::shared_ptr<Cache> block_cache = nullptr;
           std::shared_ptr<Cache> block_cache_compressed = nullptr;
+          std::map<CacheEntryRole, bool> memory_tracking_override = {};
```
Comment (hx235) [inline table/block_based/block_based_table_factory.cc:237]: NOTE: 
IIUC, this indicates that we do not support dynamic update to memory_tracking_override and serialization/de-serialization. I am not quite sure about the latter so haven't added the note to API, which I should after confirming on this.

**Comment by ajkr** [comment]
> `std::map<CacheEntryRole /* memory area */, bool /* is_tracked */> memory_tracking_override = {};`

There's a fairly high chance the future memory tracking configuration is more complex than a bool. For that reason I'd wrap the bool in a struct and put the struct in the map's value. Assuming we do that, the bool becomes inflexible because users must decide tracking even if they really only want to override some other setting. For that reason I'd make an `enum class TrackingState` with a `kFallback` as one of the enum values.

**Comment by hx235** [comment]
> > `std::map<CacheEntryRole /* memory area */, bool /* is_tracked */> memory_tracking_override = {};`
> 
> There's a fairly high chance the future memory tracking configuration is more complex than a bool. For that reason I'd wrap the bool in a struct and put the struct in the map's value. Assuming we do that, the bool becomes inflexible because users must decide tracking even if they really only want to override some other setting. For that reason I'd make an `enum class TrackingState` with a `kFallback` as one of the enum values.

Got you - let me look into that!

**Comment by pdillinger** [comment]
I think a lot of the changes here are based on an implementation distinction that I don't think the user should be exposed to or thinking about: pinned vs. tracked. For the user, things that use memory in RocksDB have one of four treatments:
* The memory usage is charged to block cache but NOT ejectable from block cache. (Internally, this could mean pinned, likely immutable, or tracked, likely mutable.)
* The memory usage is charged to block cache and IS ejectable from block cache though given HIGH priority treatment.
* The memory usage is charged to block cache and IS ejectable from block cache and given LOW priority treatment.
* The memory usage is NOT charged to block cache.

For example, if we decide to make table readers ejectable under block cache, the old state of affairs should not be "baked in" to the API.

**Comment by hx235** [comment]
Just CI - not ready for review.

File: table/block_based/block_based_table_factory.cc [hunk 2]
```diff
@@ -234,6 +234,7 @@ static std::unordered_map<std::string, OptionTypeInfo>
         /* currently not supported
           std::shared_ptr<Cache> block_cache = nullptr;
           std::shared_ptr<Cache> block_cache_compressed = nullptr;
+          CacheUsageOptions cache_usage_options;
```
Comment (hx235) [inline table/block_based/block_based_table_factory.cc:237]: NOTE:
IIUC, this indicates that we do not support dynamic update to memory_tracking_override and serialization/de-serialization. I am not quite sure about the latter so haven't added the note to API, which I should after confirming on this.

File: include/rocksdb/table.h [hunk 3]
```diff
@@ -104,6 +106,20 @@ struct MetadataCacheOptions {
...
+    kEnabled,
+    kDisabled,
+    kFallback,
+  };
+  Decision charged = Decision::kFallback;
+};
+
+struct CacheUsageOptions {
+  CacheEntryRoleOptions default_options;
+  std::array<CacheEntryRoleOptions, kNumCacheEntryRoles> options_overrides;
```
Comment (hx235) [inline include/rocksdb/table.h:120]: [NOTE]
I often find reasoning about whether data member's definition like this will be properly default-initialized (i.e, `{CacheEntryRoleOptions(), CacheEntryRoleOptions(), ..., CacheEntryRoleOptions()}`)

My reasoning for this one is that the trivial default constructor `CacheUsageOptions()` will always be called (cuz there is no other constructor) and according to the "situation 3)" in https://en.cppreference.com/w/cpp/language/default_initialization saying "when ... or a non-static data member is not mentioned in a [constructor initializer list](https://en.cppreference.com/w/cpp/language/constructor) and that constructor is called", I believe `options_overrides` will be such non-static data member and get default initialization. Then according to "if T is an array type, every element of the array is default-initialized;", I believe for each element, the trivial default constructor `CacheEntryRoleOptions()` will be called so we should be safe.

Does reasoning about things like this ever bother you? Is our CI strong enough catch undefined undetermined initialized value like this when it isn't properly default initialized?

**Comment by hx235** [comment]
Need to look into a more frequent crash_test failure  https://app.circleci.com/pipelines/github/facebook/rocksdb/13248/workflows/a5bb1b3c-f50d-4439-9ddd-d496c0f3ac70/jobs/358434 with this PR to see whether it's relevant.

**Comment by hx235** [comment]
> Need to look into a more frequent crash_test failure https://app.circleci.com/pipelines/github/facebook/rocksdb/13248/workflows/a5bb1b3c-f50d-4439-9ddd-d496c0f3ac70/jobs/358434 with this PR to see whether it's relevant.
> 
> But other than that, it's ready for review @ajkr !

Run the same command on upsteram/main with the same errors - so the db_stress failure is not related to this PR.

File: include/rocksdb/table.h (hunk 3 above)
Comment (ajkr) [reply include/rocksdb/table.h:120]: Sorry I meant to suggest std::array for the internal representation. Is there a way to still use `std::map` publicly and use an array-like representation internally? Otherwise we have to worry about preserving the value of existing enum values so introducing new CacheEntryRoles does not break existing configs. But that is difficult - one reason why is because kMisc keeps getting its value bumped as the doc says to add new enum values before it.

> Does reasoning about things like this ever bother you?

Yes, always. Since it has the intuitive behavior here it is OK with me for this point.

> Is our CI strong enough catch undefined undetermined initialized value like this when it isn't properly default initialized?

Don't think so. Last time I thought about it was here - https://github.com/facebook/rocksdb/pull/5720#pullrequestreview-276899350.

Comment (hx235) [reply include/rocksdb/table.h:120]: > Sorry I meant to suggest std::array for the internal representation. Is there a way to still use std::map publicly and use an array-like representation internally? Otherwise we have to worry about preserving the value of existing enum values so introducing new CacheEntryRoles does not break existing configs.

Aha - yeah let me fix that quickly and figure out where to store this internal rep.

File: include/rocksdb/table.h [hunk 4]
```diff
@@ -287,47 +306,88 @@ struct BlockBasedTableOptions {
...
+  // If such memory usage exceeds the avaible space left in the block cache
+  // at some point (i.e, causing a cache full under
+  // `LRUCacheOptions::strict_capacity_limit` = true), the data will then be
+  // unbuffered.
   //
-  // Charged memory usage includes:
-  // 1. Bloom Filter (format_version >= 5) and Ribbon Filter construction
-  // 2. More to come...
+  // (2) CacheEntryRole::kFilterConstruction : kDisabled
+  // If enabled, charge memory usage of Bloom Filter (format_version >= 5) and
```
Comment (ajkr) [inline include/rocksdb/table.h:366]: Since this is describing existing behavior I'd say something like "Same as kDisabled - does not charge memory usage of ...".

File: tools/db_bench_tool.cc [hunk 5]
```diff
@@ -4162,8 +4172,39 @@ class Benchmark {
...
+      block_based_options.cache_usage_options.options_overrides.insert(
+          {CacheEntryRole::kFilterConstruction,
+           {.charged = FLAGS_charge_filter_construction
+                           ? CacheEntryRoleOptions::Decision::kEnabled
+                           : CacheEntryRoleOptions::Decision::kDisabled}});
+      block_based_options.cache_usage_options.options_overrides.insert(
+          {CacheEntryRole::kFilterConstruction,
+           {.charged = FLAGS_charge_filter_construction
+                           ? CacheEntryRoleOptions::Decision::kEnabled
+                           : CacheEntryRoleOptions::Decision::kDisabled}});
```
Comment (ajkr) [inline tools/db_bench_tool.cc:4207]: These two are the same.

File: table/table_test.cc [hunk 6]
```diff
@@ -5276,6 +5294,13 @@ TEST_P(
   table_options.block_cache = cache;
   table_options.flush_block_policy_factory =
       std::make_shared<FlushBlockEveryKeyPolicyFactory>();
+  // table_options.cache_usage_options
+  //     .options_overrides[static_cast<uint32_t>(
+  //         CacheEntryRole::kCompressionDictionaryBuildingBuffer)]
+  //     .charged = CacheEntryRoleOptions::Decision::kEnabled;
+  table_options.cache_usage_options.options_overrides.insert(
+      {CacheEntryRole::kCompressionDictionaryBuildingBuffer,
+       {.charged = CacheEntryRoleOptions::Decision::kEnabled}});
```
Comment (ajkr) [inline table/table_test.cc:5303]: Is it needed? It's the default behavior to charge for it, right?

File: table/block_based/filter_policy.cc [hunk 7]
```diff
@@ -1525,9 +1537,20 @@ BloomLikeFilterPolicy::GetStandard128RibbonBuilderWithContext(
...
+  //                  CacheEntryRole::kFilterConstruction)]
+  //              .charged == CacheEntryRoleOptions::Decision::kEnabled);
+  const auto iter =
+      context.table_options.cache_usage_options.options_overrides.find(
+          CacheEntryRole::kFilterConstruction);
+  const auto& reserve_filter_construction_mem =
+      iter != context.table_options.cache_usage_options.options_overrides.end()
+          ? iter->second.charged == CacheEntryRoleOptions::Decision::kEnabled
+          : context.table_options.cache_usage_options.default_options.charged ==
+                CacheEntryRoleOptions::Decision::kEnabled;
```
Comment (ajkr) [inline table/block_based/filter_policy.cc:1553]: This expression scares me, originally because of its length/density and later because it appears to be assigning a temporary boolean to a reference. I guess the `const bool&`'s lifetime will be extended and it's indeed correct but it is too much extra difficulty to understand. Can we add an if-else and remove the reference?

File: include/rocksdb/table.h [hunk 8]
```diff
@@ -287,47 +306,88 @@ struct BlockBasedTableOptions {
   // separately
   uint64_t metadata_block_size = 4096;
 
-  // If true, a dynamically updating charge to block cache, loosely based
-  // on the actual memory usage of table building, will occur to account
-  // the memory, if block cache available.
+  // `cache_usage_options` allows users to specify the default
+  // options (`cache_usage_options.default_options`) and the overrides
+  // options (`cache_usage_options.options_overrides`)
```
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h:311]: I'll go for the second - thanks!

File: include/rocksdb/table.h [hunk 9]
```diff
@@ -287,47 +306,88 @@ struct BlockBasedTableOptions {
...
+  // is set to `Decision::kFallback`, then the existing behavior will be in
+  // effect. See "Features" below for the existing behaviors.
+  //
+  // Example:
+  // (b.1)
+  // `options_overrides[static_cast<uint32_t>(CacheEntryRole::kBlockBasedTableReader)]
+  // = {.charged = Decision::kFallback}`
+  // (b.2) `cache_usage_options.default_options.charged = Decision::kFallback`
+  // (b.1) + (b.2) => the `CacheEntryRoleOptions::charged` feature will fallback
+  // to the existing behavior for `CacheEntryRole::kBlockBasedTableReader`
```
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h:341]: The comment is basically a version of this for std::array (where we don't have "If options_overrides has an entry for role"). But my example and description might be too wordy so let me fix that with your style of listing.

File: include/rocksdb/table.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h:311]: I'll go with the first suggestion - thanks!

File: include/rocksdb/table.h (hunk 4 above)
Comment (hx235) [reply include/rocksdb/table.h:366]: Will fix!

**Comment by hx235** [comment]
Not ready for review - just going through CI

File: table/block_based/filter_policy.cc (hunk 7 above)
Comment (hx235) [reply table/block_based/filter_policy.cc]: Fixed by comparing .charged with CacheEntryRoleOptions::Decision instead of bool.

File: table/table_test.cc (hunk 6 above)
Comment (hx235) [reply table/table_test.cc]: It is not needed - it was there in place of a comment like "`CacheEntryRoleOptions::charged` is enabled by default for
 CacheEntryRole::kCompressionDictionaryBuildingBuffer" but I am removing it here as suggested.

File: db/db_sst_test.cc [hunk 10]
```diff
@@ -1497,7 +1505,20 @@ TEST_F(DBSSTTest, OpenDBWithInfiniteMaxOpenFilesSubjectToMemoryLimit) {
...
       // subject to memory limit
       Status s = TryReopen(options);
-      if (table_options.reserve_table_reader_memory) {
+      // if (table_options.cache_usage_options
+      //         .options_overrides[static_cast<uint32_t>(
+      //             CacheEntryRole::kBlockBasedTableReader)]
+      //         .charged == CacheEntryRoleOptions::Decision::kEnabled) {
+      const auto iter_1 =
+          table_options.cache_usage_options.options_overrides.find(
+              CacheEntryRole::kBlockBasedTableReader);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_sst_test.cc]: Fixed (it was chosen to quickly hack a PR to benchmark)

File: db/db_sst_test.cc [hunk 11]
```diff
@@ -1497,7 +1505,20 @@ TEST_F(DBSSTTest, OpenDBWithInfiniteMaxOpenFilesSubjectToMemoryLimit) {
       // Reopening the DB will try to load all existing files, conditionally
       // subject to memory limit
       Status s = TryReopen(options);
-      if (table_options.reserve_table_reader_memory) {
+      // if (table_options.cache_usage_options
+      //         .options_overrides[static_cast<uint32_t>(
+      //             CacheEntryRole::kBlockBasedTableReader)]
+      //         .charged == CacheEntryRoleOptions::Decision::kEnabled) {
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_sst_test.cc]: Fixed by removing the previous hacking code

File: include/rocksdb/table.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h]: Fixed

File: include/rocksdb/table.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h]: Fixed

File: include/rocksdb/table.h (hunk 4 above)
Comment (hx235) [reply include/rocksdb/table.h]: Fixed

File: tools/db_bench_tool.cc (hunk 5 above)
Comment (hx235) [reply tools/db_bench_tool.cc]: Fixed.

File: table/table_test.cc (hunk 6 above)
Comment (hx235) [reply table/table_test.cc]: Fixed

File: include/rocksdb/table.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h]: (And I realized your suggestion has the merit in " if options[role].f != kFallback, we use options[role].f" which was what I didn't have in my doc)

**Comment by hx235** [comment]
@ajkr  ready for review again!

File: table/block_based/block_based_table_factory.cc [hunk 12]
```diff
@@ -429,8 +422,16 @@ BlockBasedTableFactory::BlockBasedTableFactory(
...
   RegisterOptions(&table_options_, &block_based_table_type_info);
 
-  if (table_options_.reserve_table_reader_memory &&
-      table_options_.no_block_cache == false) {
+  const auto options_overrides_iter =
+      table_options_.cache_usage_options.options_overrides.find(
+          CacheEntryRole::kBlockBasedTableReader);
+  const auto table_reader_charged =
+      options_overrides_iter !=
+              table_options_.cache_usage_options.options_overrides.end()
```
Comment (ajkr) [inline table/block_based/block_based_table_factory.cc:430]: The InitializeOptions() logic looks like it populates entries for every `CacheEntryRole` including filling in fallbacks to default options. In that case can we use `table_options_.cache_usage_options.options_overrides.at(CacheEntryRole::kBlockBasedTableReader).charged` to get the final result directly? (Same question applies to similar code like for calculating `filter_construction_charged`.)

**Review by ajkr — APPROVED** [review]
Looks great, thanks for starting this block cache configuration API!

File: include/rocksdb/table.h [hunk 13]
```diff
@@ -287,47 +305,67 @@ struct BlockBasedTableOptions {
...
+  // back to Bloom Filter.
+  //
+  // (3) CacheEntryRole::kBlockBasedTableReader : kDisabled
+  // Same as kDisabled -  does not charge memory usage of table properties +
+  // index block/filter block/uncompression dictionary when stored in table
   // reader (i.e, BlockBasedTableOptions::cache_index_and_filter_blocks ==
   // false)
-  // 3. Some internal data structures
-  // 4. More to come...
+  // + some internal data structures during table reader creation.
```
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h]: Fixed

File: include/rocksdb/table.h [hunk 14]
```diff
@@ -287,47 +305,67 @@ struct BlockBasedTableOptions {
...
+  // If enabled and if such a table reader exceeds
+  // the avaible space left in the block cache at some point (i.e, causing
+  // a cache full under `LRUCacheOptions::strict_capacity_limit` = true),
+  // creation will fail with Status::MemoryLimit().
+  //
+  // (4) Other CacheEntryRole : Not supported
+  // `Status::kNotSupported` will be returned if
+  // `CacheEntryRoleOptions::charged` is set to {`kEnabled`, `kDisabled`}.
```
[the remark this answers is not in the document]
Comment (hx235) [reply include/rocksdb/table.h]: Fixed

File: table/block_based/block_based_table_factory.cc (hunk 12 above)
Comment (hx235) [reply table/block_based/block_based_table_factory.cc]: For `CacheEntryRole::kBlockBasedTableReader`, yeah we should remove (and already fixed it)
For `filter_construction_charged`, some tests directly initialize with FilterBitsBuilder without going through InitializeOptions like https://github.com/facebook/rocksdb/blob/7.2.fb/util/bloom_test.cc#L301 so we can't do so without fixing a lot of the tests (so I will leave it as it is)
For dictionary compression buffer, still going through CI check and see if there is any test we need to fix for that.

Comment (ajkr) [reply table/block_based/block_based_table_factory.cc]: ok with me. Sounds like an issue with the tests if they are using internal APIs in a way that could not happen in non-test code.
